keps/sig-apps/4212-declarative-node-maintenance/README.md

```diff
@@ -0,0 +1,1330 @@
+Kubernetes—i.e., [kubernetes/kubernetes], we require the following Release
+Signoff checklist to be completed.
+
+Check these off as they are completed for the Release Team to track. These
+checklist items _must_ be updated for the enhancement to be released.
+-->
+
+Items marked with (R) are required *prior to targeting to a milestone / release*.
+
+- [ ] (R) Enhancement issue in release milestone, which links to KEP dir in [kubernetes/enhancements] (not the initial KEP PR)
+- [ ] (R) KEP approvers have approved the KEP status as `implementable`
+- [ ] (R) Design details are appropriately documented
+- [ ] (R) Test plan is in place, giving consideration to SIG Architecture and SIG Testing input (including test refactors)
+  - [ ] e2e Tests for all Beta API Operations (endpoints)
+  - [ ] (R) Ensure GA e2e tests meet requirements for [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+  - [ ] (R) Minimum Two Week Window for GA e2e tests to prove flake free
+- [ ] (R) Graduation criteria is in place
+  - [ ] (R) [all GA Endpoints](https://github.com/kubernetes/community/pull/1806) must be hit by [Conformance Tests](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/conformance-tests.md) 
+- [ ] (R) Production readiness review completed
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+<!--
+**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
+-->
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+This KEP proposes adding a declarative API to manage a node maintenance. This API can be used to
+implement additional capabilities around node draining.
+
+## Motivation
+The goal of this KEP is to analyze and improve node maintenance in Kubernetes.
+
+Node maintenance is a request from a cluster administrator to remove all pods
+from a node(s) so that it can be disconnected from the cluster to perform a
+software upgrade (OS, Kubelet), hardware upgrade, or simply to remove the node
+as it is no longer needed.
+
+Kubernetes has existing support for this use case in the following way with `kubectl drain`:
+1. There are running pods on node A, some of which are protected with PodDisruptionBudgets (PDB).
+2. Set the node `Unschedulable` (cordon) to prevent new pods from being scheduled there.
```

Comment:
Sharing experience here (we have been doing automatic node maintenance for more than 3y, with at least 3k nodes per month in our infra):
In our first version of the automation we were using `cordon` flag and it quickly became a problem. Since users can also play with that flag we started to face conflict:
- difficulty to understand if the flag was set or not by our automation
- users changing the value of the flag (uncordon) without understanding that they were interfering with the automation

The solution was to finally use a different taint key with effect `doNotSchedule`. It turned out to be a good choice especially because you can play with multiple taint key to reflect the stage of your automation. (drainCandidate - draining - drained)

Reply:
In that story, what thing removed Pods from nodes where Pods were already executing?

Reply:
tolerations also have issues, the pods can tolerate your key or they can also tolerate everything

The advantage of cordon is that it is an existing primitive that is supported by kubectl commands. We can solve the interference problem by reconciling the field and making sure the correct value is always on the node. NodeMaintenance is a higher abstraction object, so if the users are not happy about this they should contact an admin to remove the NodeMaintenance from their node.

Nevertheless you can still go behind the scheduler's back and just set the nodeName directly.

Reply:
> In that story, what thing removed Pods from nodes

We have a dedicated controller that handle the eviction. This controller is the one dealing with thousand of nodes "drain". I put "drain" between quote because in fact the controller is not calling "kubectl drain". It does the eviction of each individual pod, and we have a lot of feature associated to control pod eviction. Some of these feature are implemented over the PDB. The PDB spec is controlled by another controller that is able to integrate signal external to kubernetes to pilot the budget. We are also able to do cross-cluster validation. ... 
If you want to know more about that I invite you to look at https://datadogon.datadoghq.com/episodes/datadog-on-kubernetes-node-management/ ... of course you can skip the intro that is meant to introduce concepts that you must already know.

> tolerations also have issues, the pods can tolerate your key or they can also tolerate everything

true, but in practice, the toleration we are talking about is not even known by application users. It is a matter of ownership.

> The advantage of cordon is that it is an existing primitive that is supported by kubectl commands

Yes this is also precisely the problem that I was pointing: existing and used, and so you will have conflict with existing scripts or users and your automation. Talking about automation we have on our side introduce a dedicated API where the user can just set a dedicated annotation that finally gives the exact same result. The controller interpret that annotation and apply the required logic (taint, drain, ...)

>  We can solve the interference problem by reconciling the field and making sure the correct value is always on the node

You can, we also had that same idea at some point. Till some users created their own controller to cordon/uncordon node during regular operation for their application. The result was a nice fight between the 2 controllers with cordon/uncordon loops on a given node (and of course associated stress on the api-server and other controllers reacting to the updates).

Reply:
> difficulty to understand if the flag was set or not by our automation

We have Server Side Apply which can help, and we (Kubernetes) could teach a tool like `kubectl` to warn if trouncing an existing field manager for `.spec.unschedulable`

Reply:
I understand that is possible to automate this in your case, but that is not a case for all k8s users today. The mission of these two KEPs is to have an extensible approach that everyone can use and hook into if needed.

I suppose, both approaches could also be used to ensure that the least amount of pods are scheduled to the targeted nodes during the conflict. We can also notify the admins via events if the conflict occurs, so that only a single node maintenance approach is used.

Reply:
updated relevant sections to solve this problem:
- https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#kubectl
- https://github.com/kubernetes/enhancements/blob/a15b78d3b4f8bc1b4a4e5b1508e4ad733e64c4fd/keps/sig-apps/4212-declarative-node-maintenance/README.md#cordon

@dbenque please open a new thread if needed